Services/source/cli/fs_utils.c

```diff
@@ -0,0 +1,330 @@
+#include <string.h>
+
+#define str(s) #s
+
+void createErrorOutput(char *pcWriteBuffer, size_t xWriteBufferLen) {
+    const char *errorMsg;
+    switch (red_errno) {
+    case 1:
+        errorMsg = "RED_EPERM";
+        break;
+    case 2:
+        errorMsg = "RED_ENOENT";
+        break;
+    case 5:
+        errorMsg = "RED_EIO";
+        break;
+    case 9:
+        errorMsg = "RED_EBADF";
+        break;
+    case 16:
+        errorMsg = "RED_EBUSY";
+        break;
+    case 17:
+        errorMsg = "RED_EEXIST";
+        break;
+    case 18:
+        errorMsg = "RED_EXDEV";
+        break;
+    case 20:
+        errorMsg = "RED_ENOTDIR";
+        break;
+    case 21:
+        errorMsg = "RED_EISDIR";
+        break;
+    case 22:
+        errorMsg = "RED_EINVAL";
+        break;
+    case 23:
+        errorMsg = "RED_ENFILE";
+        break;
+    case 24:
+        errorMsg = "RED_EMFILE";
+        break;
+    case 27:
+        errorMsg = "RED_EFBIG";
+        break;
+    case 28:
+        errorMsg = "RED_ENOSPC";
+        break;
+    case 30:
+        errorMsg = "RED_EROFS";
+        break;
+    case 31:
+        errorMsg = "RED_EMLINK";
+        break;
+    case 34:
+        errorMsg = "RED_ERANGE";
+        break;
+    case 36:
+        errorMsg = "RED_ENAMETOOLONG";
+        break;
+    case 38:
+        errorMsg = "RED_ENOSYS";
+        break;
+    case 39:
+        errorMsg = "RED_ENOTEMPTY";
+        break;
+    case 61:
+        errorMsg = "RED_ENODATA";
+        break;
+    case 87:
+        errorMsg = "RED_EUSERS";
+        break;
+    default:
+        errorMsg = "Unknown Error";
+    }
+    snprintf(pcWriteBuffer, xWriteBufferLen, "%s\n", errorMsg);
+    red_errno = 0;
+}
+
+static BaseType_t prvPWDCommand(char *pcWriteBuffer, size_t xWriteBufferLen, const char *pcCommandString) {
+    char pszBuffer[REDCONF_NAME_MAX];
+    char *cwd = red_getcwd(pszBuffer, REDCONF_NAME_MAX);
+
+    if (cwd == NULL) {
+        createErrorOutput(pcWriteBuffer, xWriteBufferLen);
+    } else {
+        snprintf(pcWriteBuffer, xWriteBufferLen, "%s\n", cwd);
+    }
+    return pdFALSE;
+}
+
+static BaseType_t prvCDCommand(char *pcWriteBuffer, size_t xWriteBufferLen, const char *pcCommandString) {
+    // We can guarantee there will be one parameter because FreeRTOS+CLI won't call this function unless it has
+    // exactly one parameter
+    BaseType_t parameterLen;
+    const char *parameter = FreeRTOS_CLIGetParameter(
+        /* The command string itself. */
+        pcCommandString,
+        /* Return the first parameter. */
+        1,
+        /* Store the parameter string length. */
+        &parameterLen);
+    int32_t error = red_chdir(parameter);
+    if (error < 0) {
+        createErrorOutput(pcWriteBuffer, xWriteBufferLen);
+    }
+    return pdFALSE;
+}
+
+static BaseType_t prvLSCommand(char *pcWriteBuffer, size_t xWriteBufferLen, const char *pcCommandString) {
+    static bool firstRun = true;
+    static REDDIR *dir;
```

Comment:
Are red_getcwd() and red_opendir() really slow? Just wondering if you're saving anything by keeping this open

Reply:
By keeping it open it saves the point we're at in the directory when continuing to read later